src/core/lib/gpr/fork.cc

```diff
@@ -69,10 +84,104 @@ void grpc_fork_support_init() {
   if (override_fork_support_enabled != -1) {
     fork_support_enabled = override_fork_support_enabled;
   }
+  if (fork_support_enabled) {
+    g_fork_state = (fork_state*)gpr_malloc(sizeof(fork_state));
+    gpr_mu_init(&g_fork_state->mu);
+    gpr_cv_init(&g_fork_state->thd_cv);
+    gpr_cv_init(&g_fork_state->exec_ctx_cv);
+    gpr_atm_no_barrier_store(&g_fork_state->exec_ctx_count, 2);
+    g_fork_state->thd_count = 0;
+    g_fork_state->awaiting_threads = false;
+  }
+}
+
+void grpc_fork_support_destroy() {
+  if (g_fork_state) {
+    gpr_mu_destroy(&g_fork_state->mu);
+    gpr_cv_destroy(&g_fork_state->thd_cv);
+    gpr_cv_destroy(&g_fork_state->exec_ctx_cv);
+    gpr_free(g_fork_state);
+    g_fork_state = nullptr;
+  }
+}
+
+void grpc_fork_inc_thd_count() {
+  if (g_fork_state) {
+    gpr_mu_lock(&g_fork_state->mu);
+    g_fork_state->thd_count++;
+    gpr_mu_unlock(&g_fork_state->mu);
+  }
+}
+
+void grpc_fork_dec_thd_count() {
+  if (g_fork_state) {
+    gpr_mu_lock(&g_fork_state->mu);
+    g_fork_state->thd_count--;
+    if (g_fork_state->awaiting_threads && g_fork_state->thd_count == 0) {
+      gpr_cv_signal(&g_fork_state->thd_cv);
+    }
+    gpr_mu_unlock(&g_fork_state->mu);
+  }
+}
+
+void grpc_fork_await_thds() {
+  if (g_fork_state) {
+    gpr_mu_lock(&g_fork_state->mu);
+    g_fork_state->awaiting_threads = true;
+    gpr_cv_wait(&g_fork_state->thd_cv, &g_fork_state->mu,
+                gpr_inf_future(GPR_CLOCK_REALTIME));
+    g_fork_state->awaiting_threads = false;
+    gpr_mu_unlock(&g_fork_state->mu);
+  }
+}
+
+void grpc_fork_inc_exec_ctx_count() {
+  if (g_fork_state) {
+    intptr_t count = gpr_atm_no_barrier_load(&g_fork_state->exec_ctx_count);
+    while (true) {
+      if (count <= 1) {
+        // This only occurs if we are trying to fork.  Wait until the fork()
+        // operation completes before allowing new ExecCtxs.
+        gpr_mu_lock(&g_fork_state->mu);
+        if (gpr_atm_no_barrier_load(&g_fork_state->exec_ctx_count) <= 1) {
+          gpr_cv_wait(&g_fork_state->exec_ctx_cv, &g_fork_state->mu,
+                      gpr_inf_future(GPR_CLOCK_REALTIME));
+        }
+        gpr_mu_unlock(&g_fork_state->mu);
+      } else if (gpr_atm_no_barrier_cas(&g_fork_state->exec_ctx_count, count,
+                                        count + 1)) {
+        break;
+      }
+      count = gpr_atm_no_barrier_load(&g_fork_state->exec_ctx_count);
+    }
+  }
+}
+
+void grpc_fork_dec_exec_ctx_count() {
+  if (g_fork_state) {
+    gpr_atm_no_barrier_fetch_add(&g_fork_state->exec_ctx_count, -1);
+  }
+}
+
+bool grpc_fork_block_exec_ctx() {
+  if (g_fork_state) {
+    // Assumes there is an active ExecCtx when this function is called
+    return gpr_atm_no_barrier_cas(&g_fork_state->exec_ctx_count, 3, 1);
```

Comment:
3 is a funny magic number. Let's not have magic numbers in the code.

Reply:
I've added some macros and a comment describing how these numbers are formulated.